arrows/ocv/three_frame_differencing.cxx

```diff
@@ -0,0 +1,435 @@
+/*ckwg +29
+ * Copyright 2017 by Kitware, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * \file
+ * \brief Implementation of ocv::three_frame_differencing
+ */
+
+#include <deque>
+
+#include "three_frame_differencing.h"
+
+#include <kwiversys/SystemTools.hxx>
+#include <vital/exceptions.h>
+
+#include <arrows/ocv/image_container.h>
+
+#include <opencv2/opencv.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+#include <opencv2/core/eigen.hpp>
+
+namespace kwiver {
+namespace arrows {
+namespace ocv {
+
+using namespace kwiver::vital;
+
+
+//-----------------------------------------------------------------------------
+  ///
+  /**
+  * @brief Converts a multi-channel image into a single channel image
+  *
+  * Replace multi-channel image with a single channel image equal to the root
+  * mean square over the channels.
+  *
+  * @param src first image
+  * @param dst second image
+  * @param dst_type OpenCV data type
+  */
+static
+void
+rms_over_channels( const cv::Mat &src, cv::Mat &dst, int dst_type)
+{
+  cv::Mat src_split[src.channels()];
+  cv::split(src, src_split);
+  cv::Mat accum = cv::Mat(src.rows, src.cols, CV_32F, cvScalar(0));
+  for( int i=0; i<src.channels(); ++i)
+  {
+    cv::Mat temp;
+    src_split[i].convertTo(temp, CV_32F);
+
+    // Divide the result by 3^2 so that the difference has the same scale as
+    // a mono image
+    cv::multiply( temp, temp, temp, 1/9.0 );
```

Comment:
I think the 9 here should be a 3.  Duplicating these operations in equivalent numpy does not yield correct values:
```python
>>> img = np.ndarray((3, 3, 3), dtype=float)
>>> img[:] = 255
>>> # Same as current implementation
>>> np.sqrt((img * img / 9.).sum(axis=2))
array([[ 147.22431864,  147.22431864,  147.22431864],
       [ 147.22431864,  147.22431864,  147.22431864],
       [ 147.22431864,  147.22431864,  147.22431864]])
>>> # With 9 -> 3
>>> np.sqrt((img * img / 3.).sum(axis=2))
array([[ 255.,  255.,  255.],
       [ 255.,  255.,  255.],
       [ 255.,  255.,  255.]])
```

PTAL @mattbrown11

Reply:
Yeah, should be 

`cv::multiply( temp, temp, temp, 1/3.0 );`